pkg/scheduler/factory.go

```diff
@@ -268,138 +267,56 @@ func (c *Configurator) createFromConfig(policy schedulerapi.Policy) (*Scheduler,
 
-	pluginsForPredicates, pluginConfigForPredicates, err := getPredicateConfigs(predicateKeys, lr, args)
-	if err != nil {
-		return nil, err
-	}
-
-	pluginsForPriorities, pluginConfigForPriorities, err := getPriorityConfigs(priorityKeys, lr, args)
-	if err != nil {
-		return nil, err
-	}
 	// Combine all framework configurations. If this results in any duplication, framework
 	// instantiation should fail.
-	var defPlugins schedulerapi.Plugins
+
 	// "PrioritySort" and "DefaultBinder" were neither predicates nor priorities
 	// before. We add them by default.
-	defPlugins.Append(&schedulerapi.Plugins{
+	plugins := schedulerapi.Plugins{
 		QueueSort: &schedulerapi.PluginSet{
 			Enabled: []schedulerapi.Plugin{{Name: queuesort.Name}},
 		},
 		Bind: &schedulerapi.PluginSet{
 			Enabled: []schedulerapi.Plugin{{Name: defaultbinder.Name}},
 		},
-	})
-	defPlugins.Append(pluginsForPredicates)
-	defPlugins.Append(pluginsForPriorities)
-	defPluginConfig, err := mergePluginConfigsFromPolicy(pluginConfigForPredicates, pluginConfigForPriorities)
-	if err != nil {
+	}
+	var pluginConfig []schedulerapi.PluginConfig
+	var err error
+	if plugins, pluginConfig, err = lr.AppendPredicateConfigs(predicateKeys, args, plugins, pluginConfig); err != nil {
+		return nil, err
+	}
+	if plugins, pluginConfig, err = lr.AppendPriorityConfigs(priorityKeys, args, plugins, pluginConfig); err != nil {
+		return nil, err
+	}
+	if pluginConfig, err = dedupPluginConfigs(pluginConfig); err != nil {
 		return nil, err
 	}
 	for i := range c.profiles {
 		prof := &c.profiles[i]
-		// Plugins are empty when using Policy.
+		// Plugins and PluginConfig are empty when using Policy; overriding.
 		prof.Plugins = &schedulerapi.Plugins{}
-		prof.Plugins.Append(&defPlugins)
-
-		// PluginConfig is ignored when using Policy.
-		prof.PluginConfig = defPluginConfig
+		prof.Plugins.Append(&plugins)
+		prof.PluginConfig = pluginConfig
 	}
 
 	return c.create()
 }
 
-// mergePluginConfigsFromPolicy merges the giving plugin configs ensuring that,
+// dedupPluginConfigs removes duplicates from pluginConfig, ensuring that,
 // if a plugin name is repeated, the arguments are the same.
-func mergePluginConfigsFromPolicy(pc1, pc2 []schedulerapi.PluginConfig) ([]schedulerapi.PluginConfig, error) {
+func dedupPluginConfigs(pc []schedulerapi.PluginConfig) ([]schedulerapi.PluginConfig, error) {
 	args := make(map[string]runtime.Object)
-	for _, c := range pc1 {
-		args[c.Name] = c.Args
-	}
-	for _, c := range pc2 {
-		if v, ok := args[c.Name]; ok && !cmp.Equal(v, c.Args) {
+	result := make([]schedulerapi.PluginConfig, 0, len(pc))
+	for _, c := range pc {
+		if v, found := args[c.Name]; !found {
+			result = append(result, c)
+		} else if !cmp.Equal(v, c.Args) {
 			// This should be unreachable.
 			return nil, fmt.Errorf("inconsistent configuration produced for plugin %s", c.Name)
 		}
 		args[c.Name] = c.Args
```

Comment:
nit: move it to L312; otherwise, this line is called N-1 times upon every N duplicate configs